### ISSUE TITLE
Process always crashes when closing WebSocket on the server side

### ISSUE THREAD

**Issue by ygoe** [dotnet/aspnetcore#24096]
### Describe the bug
Our application is a client-server application where clients connect to the server through WebSocket. Those connections are held by the client during its entire runtime. When the server shuts down, those connections must be closed by the server. This is triggered by the `IHostApplicationLifetime.ApplicationStopping` cancellation token. While the server handles an open connection from a client in its controller action method, it has a registration on that token that then calls `WebSocket.CloseAsync`.

This has always worked fine in .NET Core 2.2 but fails miserably in .NET Core 3.1. After upgrading the server application, when it should shut down, it crashes instead. Different output is generated on the console each time and the output is not reproducible. But the crash happens every time, the process never ends normally. All code paths have multiple try/catch handlers around them and all async calls are awaited but nothing helps. This web socket throws an uncatchable exception. Sometimes, there is no console output at all (before systemd says the process has aborted), sometimes I see slightly chaotic output like this:

    Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. fail: AppName.Services.RmpRouter[0]

It looks like the "fail:" starts another last log message and the repeated "Unhandled exception." before have no line termination. And sometimes it writes out a bit more before it dies:

    Unhandled exception.Unhandled exception.Unhandled exception.   Unhandled exception. System.OperationCanceledException: The read was canceled
       at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)

I'm not sure this is all it wanted to write, it probably died midway.

In rare cases, the exception(s?) made it further into my catch blocks and was logged partially. It seems that the process crashed while it was still writing its log lines to the console (where they are picked up by systemd), so the log is unusable in this case because it's not reliable. I have now added more direct `Console.WriteLine` calls but most of the time, nothing is printed at all. The process just disappears. That's not why I'm using .NET – I'm used to get reliable exception traces instead of the typical silent crashes of an unmanaged C application.

I believe that the unhandled exception is thrown in code that's out of my scope. It must be somewhere in the runtime/framework classes and there's a try/catch handler missing somewhere. And it was introduced between .NET Core 2.2 and 3.1. And it's kind of a critical issue because we can't cleanly shut down our server application anymore now, leading to data loss in unpredictable forms.

### To Reproduce
Sorry, nothing to share here. The application is too big (and non-public) and I don't know how I should create a copy of it that still shows the bug. There's also an additional NuGet package of mine involved in handling those connections. It would require a large effort in the range of days to set this up.

I could try to analyse this further if I knew where to start. I've tried all I could think of, the proces just dies when it really shouldn't. That's beyond my capabilities now.

### Exceptions (if any)
Certainly, but nothing really much to see of it. See above.

### Further technical details
- ASP.NET Core version: 3.1.6
- Running on Ubuntu 16.04 x64

**Comment by davidfowl** [comment]
> Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. fail: AppName.Services.RmpRouter[0]

Do you have any more content? Or is it actually what gets printed verbatim without edits?

> This has always worked fine in .NET Core 2.2 but fails miserably in .NET Core 3.1

It is safe to assume you were always using Kestrel?

> Unhandled exception.Unhandled exception.Unhandled exception.   Unhandled exception. System.OperationCanceledException: The read was canceled
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)

Is this the full stack trace for the exception? 

Do you have any async void in your code base? One common problem with websockets is bad handling of the lifetime, yielding the request before the websocket is torn down (https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets?view=aspnetcore-3.1#accept-websocket-requests)

**Comment by ygoe** [comment]
I have the following output from several runs. It is slightly filtered from journald, removing the host and service name. The lines marked with `*` after the timestamp are generated by my application. The web application should still print a log message and this first block shows what it used to be with ASP.NET Core 2.2:

```
Jul 18 12:01:54 systemd[1]: Stopping aspnetapp-​7000.service...
Jul 18 12:01:54*Application is shutting down...
Jul 18 12:01:58 systemd[1]: Stopped aspnetapp-​​7000.service.
```

This was the first bad occurrence with .NET Core 3.1: (The final ABRT line occurs every time.)
```
Jul 19 15:45:42 Unhandled exception. Unhandled exception. Unhandled exception. fail: MyApp.Services.RmpRouter[0]
Jul 19 15:47:12 systemd[1]: aspnetapp-​7000.service: State 'stop-​sigterm' timed out. Killing.
Jul 19 15:47:12 systemd[1]: aspnetapp-​​7000.service: Main process exited, code=killed, status=6/​ABRT
Jul 19 15:47:12 systemd[1]: Stopped aspnetapp-​7000.service.
```

```
Jul 19 15:49:22 Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. fail: MyApp.Services.RmpRouter[0]
Jul 19 15:49:22*      The read was canceled (user=... ip=...)
Jul 19 15:49:22 System.OperationCanceledException: The read was canceled
Jul 19 15:49:22    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
Jul 19 15:49:22    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)
```

```
Jul 19 16:04:56 Unhandled exception.Unhandled exception.  Unhandled exception. fail: MyApp.Services.RmpRouter[0]
Jul 19 16:04:56*      The read was canceled (user=... ip=...)
```

```
Jul 19 16:05:40 Unhandled exception.Unhandled exception.  Unhandled exception. fail: MyApp.Services.RmpRouter[0]
```

Here is the only occurrence with more output. At least two clients were connected at the time (user1, user2). The exceptions were logged in parallel which overcharged the logger.
```
Jul 19 16:06:57 Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception. fail: MyApp.Services.RmpRouter[0]
Jul 19 16:06:57*      The read was canceled (user=user1 ip=...)
Jul 19 16:06:57 System.OperationCanceledException: The read was canceled
Jul 19 16:06:57    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.CloseAsyncPrivate(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.CloseAsync(WebSocket webSocket, SemaphoreSlim wsSendMutex, WebSocketCloseStatus status, String message)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.<>c__DisplayClass48_0.<<HandleClientConnection>b__3>d.MoveNext()
Jul 19 16:06:57 System.OperationCanceledException: The read was canceled
Jul 19 16:06:57    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.CloseAsyncPrivate(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.CloseAsync(WebSocket webSocket, SemaphoreSlim wsSendMutex, WebSocketCloseStatus status, String message)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.<>c__DisplayClass48_0.<<HandleClientConnection>b__3>d.MoveNext()
Jul 19 16:06:57 -​-​-​ End of stack trace from previous location where exception was thrown -​-​-​
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.ReceiveRawMessage(ClientConnectionInfo client, WebSocket webSocket)
Jul 19 16:06:57    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
Jul 19 16:06:57    at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
Jul 19 16:06:57    at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState&​ state)
Jul 19 16:06:57    at System.Threading.QueueUserWorkItemCallback.Execute()
Jul 19 16:06:57    at System.Threading.ThreadPoolWorkQueue.Dispatch()
Jul 19 16:06:57 -​-​-​ End of stack trace from previous location where exception was thrown -​-​-​
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.HandleClientConnection(IPAddress remoteIpAddress, WebSocket webSocket, ClaimsPrincipal claimsPrincipal, CancellationToken cancellationToken)
Jul 19 16:06:57*fail: MyApp.Services.RmpRouter[0]
Jul 19 16:06:57*      The read was canceled (user=user2 ip=...)
Jul 19 16:06:57 System.OperationCanceledException: The read was canceled
Jul 19 16:06:57    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)
Jul 19 16:06:57    at System.Net.WebSockets.ManagedWebSocket.CloseAsyncPrivate(WebSocketCloseStatus closeStatus, String statusDescription, CancellationToken cancellationToken)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.CloseAsync(WebSocket webSocket, SemaphoreSlim wsSendMutex, WebSocketCloseStatus status, String message)
Jul 19 16:06:57    at RoutedMessagingProtocol.Router.<>c__DisplayClass48_0.<<HandleClientConnection>b__3>d.MoveNext()
Jul 19 16:06:57 -​-​-​ End of stack trace from previous location where exception was thrown -​-​-​
```

```
Jul 19 16:43:58 Unhandled exception.Unhandled exception.Unhandled exception.   Unhandled exception. System.OperationCanceledException: The read was canceled
Jul 19 16:43:58    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
```

The more `Console.WriteLine` logging I added, the shorter the output became. Now when I try it, most of the time there is no output at all!

> It is safe to assume you were always using Kestrel?

Yes, sorry, I didn't think of there might still be other options. The web app is started via systemd and accessed through a reverse proxy through Apache which also adds SSL.

> Is this the full stack trace for the exception?

Probably not. The logging cannot complete before the process dies. That's because I don't see what and where the real issue is.

> Do you have any async void in your code base?

There are some of them but all have a try/catch around their entire body and none of them should be called in this scenario. Everything else returns a `Task`. The controller looks like this:

```c#
[Route("api/[controller]")]
public class RmpController : Controller
{
	private readonly RmpRouter router;

	public async Task Index()
	{
		if (HttpContext.WebSockets.IsWebSocketRequest)
		{
			var webSocket = await HttpContext.WebSockets.AcceptWebSocketAsync();
			logger.LogDebug($"RMP client connection from {HttpContext.Connection.RemoteIpAddress} opened");
			try
			{
				await router.HandleClientConnection(HttpContext.Connection.RemoteIpAddress, webSocket, HttpContext.User, applicationLifetime.ApplicationStopping);
			}
			catch (Exception ex)
			{
				Console.WriteLine("Log (RmpController.Index): " + ex.ToString());
				logger.LogError(ex, "RMP error: " + ex.GetRecursiveMessage());
			}
			finally
			{
				logger.LogDebug($"RMP client connection from {HttpContext.Connection.RemoteIpAddress} closed");
			}
		}
		else
		{
			HttpContext.Response.StatusCode = 400;
		}
	}
}
```

A rough simplification of the router's code is this. The router class handles a client websocket connection, receives data from it and works with it. If the router is shut down, the connection is closed by the router. This is when the trouble begins.

```c#
public async Task HandleClientConnection(IPAddress remoteIpAddress, WebSocket webSocket, ClaimsPrincipal claimsPrincipal, CancellationToken cancellationToken = default)
{
	try
	{
		cancellationTokenRegistration = cancellationToken.Register(async () =>
		{
			Console.WriteLine("Log (Router.HandleClientConnection): cancellationToken triggered, closing client WebSocket for user " + client.AuthenticatedUserName);
			try
			{
				await CloseAsync(client.WebSocket, client.SendMutex, WebSocketCloseStatus.EndpointUnavailable, "Shutting down");
			}
			catch (Exception ex)
			{
				Console.WriteLine("Log (Router.HandleClientConnection): " + ex);
			}
			Console.WriteLine("Log (Router.HandleClientConnection): cancellationToken triggered, client WebSocket for user " + client.AuthenticatedUserName + " closed");
		});

		// Message loop
		while (true)
		{
			var segment = await ReceiveRawMessage(client, webSocket);
			//...
		}
	}
	catch (OperationCanceledException ex) when (ex.CancellationToken == shutdownCts.Token || ex.CancellationToken == cancellationToken)
	{
		Console.WriteLine("Log (Router.HandleClientConnection 1): " + ex.ToString());
		OnClientError(client, "Shutting down", null);
		await CloseAsync(webSocket, client.SendMutex, WebSocketCloseStatus.EndpointUnavailable, "Shutting down");
	}
	catch (OperationCanceledException ex) when (ex.CancellationToken == client.AbortReceiveCts.Token)
	{
		Console.WriteLine("Log (Router.HandleClientConnection 2): " + ex.ToString());
		OnClientError(client, "Connection aborted", null);
		await CloseAsync(webSocket, client.SendMutex, WebSocketCloseStatus.EndpointUnavailable, "Connection aborted");
	}
	catch (Exception ex)
	{
		Console.WriteLine("Log (Router.HandleClientConnection 3): " + ex.ToString());
		OnClientError(client, ex.Message, ex);
		await CloseAsync(webSocket, client.SendMutex, WebSocketCloseStatus.InternalServerError, ex.Message);
	}
	finally
	{
		cancellationTokenRegistration.Dispose();
		try
		{
			await CloseAsync(webSocket, client.SendMutex, WebSocketCloseStatus.NormalClosure, "");
		}
		catch (Exception ex)
		{
			Console.WriteLine("Log (Router.HandleClientConnection finally): " + ex.ToString());
			OnClientError(client, ex.Message, ex);
		}
	}
}

private async Task<ArraySegment<byte>> ReceiveRawMessage(ClientConnectionInfo client, WebSocket webSocket)
{
	byte[] buffer = new byte[1024];
	int usedBufferCount = 0;
	while (true)
	{
		var segment = new ArraySegment<byte>(buffer, usedBufferCount, buffer.Length - usedBufferCount);
		var result = await webSocket.ReceiveAsync(segment, client.AbortReceiveCts.Token);
		usedBufferCount += result.Count;
		if (result.CloseStatus != null)
		{
			return new ArraySegment<byte>();
		}
		if (result.EndOfMessage)
		{
			break;
		}
		if (usedBufferCount > buffer.Length - 128)
		{
			Array.Resize(ref buffer, buffer.Length * 2);
		}
	}
	return new ArraySegment<byte>(buffer, 0, usedBufferCount);
}
```

**Comment by ygoe** [comment]
I have decreased the log level from Information to Debug and see some additional entries now: (see the code above for where they're written)

```
Jul 20 10:54:01 Unhandled exception. Unhandled exception. Log (RmpRouter.OnClientError): The read was canceled (user=... ip=...)
Jul 20 10:54:01*Log (RmpRouter.OnClientError): The read was canceled (user=user1 ip=clientip1...)
Jul 20 10:54:01*Log (RmpRouter.OnClientError): The read was canceled (user=user2 ip=clientip2...)
Jul 20 10:54:01 Unhandled exception. dbug: MyApp.Controllers.RmpController[0]
Jul 20 10:54:01*      RMP client connection from ip1... closed
Jul 20 10:54:01*dbug: MyApp.Controllers.RmpController[0]
Jul 20 10:54:01*      RMP client connection from ip2... closed
Jul 20 10:54:01*Log (RmpRouter.OnClientError): System.OperationCanceledException: The read was canceled
Jul 20 10:54:01 systemd[1]: aspnetapp-​7000.service: Main process exited, code=killed, status=6/​ABRT
```

So the API controller sees the connection as closed and the call to `HandleClientConnection` came back without an exception on this call path. The errors start somewhere else.

**Comment by ygoe** [comment]
I discovered that I didn't deploy the new file of the NuGet package DLL (I'm copying that manually now because I constantly make changes to it and see their result). Now that I use the current build of that file, the situation has changed slightly:

The process never aborts in a way that systemd would notice it. So it looks like the exit code is always 0. The messages like "Unhandled exception." are also gone completely. Maybe I missed an exception handler before that is now added.

But it's still not all good. I've added much more logging to see when all `IHostedService` instances have properly shut down with their `StopAsync` method, and when the application is finished completely in Program.cs:

```c#
public static void Main(string[] args)
{
	Console.WriteLine("MyApp is starting...");
	CreateHostBuilder(args).Build().Run();
	Console.WriteLine("MyApp is shut down");
}
```

Sometimes I see all the shutdown messages I expect, until the final "MyApp is shut down". And sometimes, it still stops midway. Then none of the `IHostedService`s are shut down and the final message is also missing.

So now the situation is: It doesn't fail always, only sometimes. And when it does, it fails absolutely silently, with no output or exit code. I'm not sure this is an improvement.

**Comment by davidfowl** [comment]
For the original error: You have an async void in there that *might* be problematic:

```C#
cancellationTokenRegistration = cancellationToken.Register(async () =>
		{
			Console.WriteLine("Log (Router.HandleClientConnection): cancellationToken triggered, closing client WebSocket for user " + client.AuthenticatedUserName);
			try
			{
				await CloseAsync(client.WebSocket, client.SendMutex, WebSocketCloseStatus.EndpointUnavailable, "Shutting down");
			}
			catch (Exception ex)
			{
				Console.WriteLine("Log (Router.HandleClientConnection): " + ex);
			}
			Console.WriteLine("Log (Router.HandleClientConnection): cancellationToken triggered, client WebSocket for user " + client.AuthenticatedUserName + " closed");
		});
```

Don't run any async code in these callbacks as nobody will be waiting on them. I would instead use the token to end the current async execution and run this logic in the normal place where you already have a call to CloseAsync.

> So now the situation is: It doesn't fail always, only sometimes. And when it does, it fails absolutely silently, with no output or exit code. I'm not sure this is an improvement.

So it seems like there's no crash and we're discussing a different kind of error? What's the reproduction for this one?

**Comment by ygoe** [comment]
> Don't run any async code in these callbacks as nobody will be waiting on them.

I've changed the async/await with `.GetAwaiter().GetResult()` for now. Now the "closing client WebSocket" and "client WebSocket closed" messages appear closer to each other and mostly for one client after another. Would this synchronous await style have any effect here?

> I would instead use the token to end the current async execution and run this logic in the normal place where you already have a call to CloseAsync.

Do you mean I should pass the cancellation token into `WebSocket.ReceiveAsync()`? I'm going to try that.

Meanwhile I've added other types of logging, `AppDomain.FirstChanceException` and `AppDomain.UnhandledException`, writing to the console. While the first is working, the second is not. The process still just disappears from the world without a sound. There are several code instructions in place that make this impossible. There must be an error somewhere in the runtime.

> What's the reproduction for this one?

Still the same: Stop the application with systemd. I have now added the systemd hosting support but that doesn't change anything.

**Comment by ygoe** [comment]
I have changed the WebSocket receive cancellation so that it only uses a cancellation token (linked with the one that was already there) instead of closing the socket from the registered handler. This has improved the exceptions (much less exceptions flying around). I'm not sure why I haven't done that earlier.

But the process still doesn't always shut down properly. Most of the time, I see the log messages of when the WebSocket controller action has returned from the connection. Often most of the `IHostedService`s are also properly stopped. Sometimes, the process disappears before all services are stopped, and sometimes, it disappears immediately with no output whatsoever after the systemd stop command. It tends to end prematurely more often when the process was running for a while (15+ minutes).

For testing, I've now disabled WebSocket support in the application by removing the following code from `Startup.Configure`:

```c#
app.UseWebSockets(new WebSocketOptions
{
	KeepAliveInterval = TimeSpan.FromSeconds(120)
});
```

Now there were no more clients connected through WebSocket. With this modification, I could still reproduce the premature exit. So I guess the problem is not from WebSocket, it's from something else. WebSocket just took time so that less of the other work was done. In fact the time seems to be restricted. I added this test code to the `Main` method:

```c#
try
{
	Console.WriteLine("MyApp is starting...");
	CreateHostBuilder(args).Build().Run();
	Console.WriteLine("MyApp is shut down");
	for (int i = 1; i <= 20; i++)
	{
		Thread.Sleep(10);
		Console.WriteLine("waited " + (i * 10) + " ms");
	}
	return 0;
}
catch (Exception ex)
{
	Console.WriteLine("Unhandled exception in Main: " + ex);
	return 1;
}
```

The result at the first time was this:

```
MyApp is shut down
waited 10 ms
waited 20 ms
```

It wanted to count further but wasn't let to. All other times none of the wait messages appeared, sometimes not even the shutdown and some of the service stop messages before that.

I've also checked all async delegates in the entire project. Some of them were just unawaited but had their own try/catch in them. I've replaced them with a `Task.Forget()` extension method that awaits the task and catches any exceptions. So if `await` ensures that the task continuation is always executed, then there are no more unawaited tasks.

I'm most struggling with how the process can end with exit code 0 even if it clearly must print some more messages, at least the one in the `Main` method. This can only be possible if something like `[TerminateProcess](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess)` (or its Linux equivalent) is called on the current process. Does (ASP).NET Core do that?
